### PR TITLE
Revise never type fallback algorithm

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This is a rebase of https://github.com/rust-lang/rust/pull/84573, but dropping the stabilization of never type (and the accompanying large test diff).

Each commit builds & has tests updated alongside it, and could be reviewed in a more or less standalone fashion. But it may make more sense to review the PR as a whole, I'm not sure. It should be noted that tests being updated isn't really a good indicator of final behavior -- never_type_fallback is not enabled by default in this PR, so we can't really see the full effects of the commits here.

This combines the work by Niko, which is [documented in this gist](https://gist.github.com/nikomatsakis/7a07b265dc12f5c3b3bd0422018fa660), with some additional rules largely derived to target specific known patterns that regress with the algorithm solely derived by Niko. We build these from an intuition that:

* In general, fallback to `()` is *sound* in all cases
* But, in general, we *prefer* fallback to `!` as it accepts more code, particularly that written to intentionally use `!` (e.g., Result's with a Infallible/! variant).

When evaluating Niko's proposed algorithm, we find that there are certain cases where fallback to `!` leads to compilation failures in real-world code, and fallback to `()` fixes those errors. In order to allow for stabilization, we need to fix a good portion of these patterns.

The final rule set this PR proposes is that, by default, we fallback from `?T` to `!`, with the following exceptions:

1. `?T: Foo` and `Bar::Baz = ?T` and `(): Foo`, then fallback to `()`
2. Per [Niko's algorithm](https://gist.github.com/nikomatsakis/7a07b265dc12f5c3b3bd0422018fa660#proposal-fallback-chooses-between--and--based-on-the-coercion-graph), the "live" `?T` also fallback to `()`.

The first rule is necessary to address a fairly common pattern which boils down to something like the snippet below. Without rule 1, we do not see the closure's return type as needing a () fallback, which leads to compilation failure.

```rust
#![feature(never_type_fallback)]

trait Bar { }
impl Bar for () {  }
impl Bar for u32 {  }

fn foo<R: Bar>(_: impl Fn() -> R) {}

fn main() {
    foo(|| panic!());
}
```

r? @jackh726

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
@jackh726 The opening comment is hopefully helpful, but in terms of review I feel like the best bet is for you to take a look at the PR as a whole at a high level, calling out notable deficiencies or questions before we dive in to each commit in detail. But really up to you.

At a high level, this PR has the first several commits which add Niko's coercion graph and detect when we need to fallback to (), and then the one commit from me atop that which includes the special case and code necessary to handle it. The coercion graph is fairly self explanatory in some sense in terms of the code, though is likely not very intuitive in terms of the mapping to effect on actual Rust code. I'm not sure I fully grasp it at a low level - mostly because coercion and subtyping predicates aren't (to me) obviously visible in Rust source code.

My special casing instruments predicate registration to capture the appropriate projection and trait predicates and stores away the inference variables. I'm not confident in this code being placed appropriately -- happy to move it around. My main goal was to avoid a dependency on `select_obligations_where_possible` not being called "before" us and eliminating the obligations we need to successfully determine which direction we should fall back. That said, it's possible I got some things wrong there.

There's also a few commented sections and FIXMEs, some of which likely could use a more experienced hand to help with -- but I don't think they represent major roadblocks, more like polish, so we can iterate on that after the high level direction I suspect.

----

FWIW, I believe this PR shouldn't affect end-user visible behavior modulo new ICEs or something (not that I'm aware of any). So it can be merged once we deem it ready, and then we'd follow up with a stabilization PR and testing. My impression based on discussions with @nikomatsakis is that never type is sufficiently unstable that iterating like this on nightly is OK -- we will want to poll T-lang for opinions on the algorithm as part of a stabilization report, of course, but there is no need to surface before then. Niko's version of the algorithm was already discussed in T-lang as well, and my addition is a pretty small delta anyway.

**Comment by Mark-Simulacrum** [comment]
(Rebased to drop first commit, now that the PR has been merged).

File: compiler/rustc_typeck/src/check/fallback.rs
Comment (jackh726) [inline compiler/rustc_typeck/src/check/fallback.rs:418]: I think this needs a comment...

Is the idea that even though it's not a "coercion", it still counts because we effectively unify them?

File: compiler/rustc_typeck/src/check/fallback.rs
Comment (jackh726) [inline compiler/rustc_typeck/src/check/fallback.rs:430]: Does this need to be `pub`? I'm now seeing where it gets used.

File: src/test/ui/never_type/fallback-closure-wrap.rs
Comment (jackh726) [inline src/test/ui/never_type/fallback-closure-wrap.rs:1]: Can you add a comment for this test?

And add the example in the OP (for projection outputs) as a test?

File: compiler/rustc_typeck/src/check/fallback.rs
Comment (jackh726) [inline compiler/rustc_typeck/src/check/fallback.rs:378]: Please add a comment here about why this is important, relevant test, etc.

**Review by jackh726 — CHANGES_REQUESTED** [review]
A few comments. Most of the commits look good. I have some reservations about the last commit, as discussed on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/259160-t-lang.2Fproject-never-type/topic/PR.20.2388804). But, I've reviewed based on "the logic here is what we want, at least for now" and "there might be another way to do this, but this is straightforward enough"

**Comment by jackh726** [comment]
Going to queue a perf run. That last commit looks suspicious.

@bors try @rust-timer queue

**Comment by jackh726** [comment]
Wow! Surprising...did I miss something.

File: compiler/rustc_trait_selection/src/traits/relationships.rs
Comment (Mark-Simulacrum) [reply compiler/rustc_trait_selection/src/traits/relationships.rs:18]: Ah, dropped this comment now. That function doesn't take the right parameters, so it can't preserve constness for example -- there might be a refactoring of that function in order, not sure. It's also private to the error_reporting module...

File: compiler/rustc_trait_selection/src/traits/relationships.rs
Comment (Mark-Simulacrum) [reply compiler/rustc_trait_selection/src/traits/relationships.rs:86]: Yeah, this is a bit complicated. I think it's probably fine to just delete the else branch here. Unfortunately, that does mean that some cases won't be considered, I think -- for example, if you register `<ConcreteType as Trait>::Proj = Foo`, that will not end up registering as 'live' for any of the involved types, even if we eventually unify `<ConcreteType as Trait>::Proj` with `?T`, I think, so we may miss that we wanted to fall back to () in this case.

But it should be sound, still, so I don't think this is actually a problem, just 'less nice', in some sesne.

File: compiler/rustc_trait_selection/src/traits/relationships.rs
Comment (nikomatsakis) [reply compiler/rustc_trait_selection/src/traits/relationships.rs:43]: I had the same thought. It seems like we should only do this if we observe that some potentially diverging variable is involved. Is there a reason that doesn't work? (Maybe this was covered on zulip?)

File: compiler/rustc_middle/src/ty/mod.rs
Comment (nikomatsakis) [inline compiler/rustc_middle/src/ty/mod.rs:2101]: We should document *why* these conditions are being used -- but it would be fine (imo) for this to be on a never-type-initiative repo (and linked from here).

**Comment by Mark-Simulacrum** [comment]
Ok, pushed up a rebase that actually ended up only touching the last commit (last two, now). Going to follow up to several comments here:

> We should document why these conditions are being used -- but it would be fine (imo) for this to be on a never-type-initiative repo (and linked from here).

I am not sure that we will end up with "great" reasoning beyond the 'it fixes regressions'. In discussions we've had about the fallback rules we *could* propose, I believe we largely ended up concluding that there's not really a rule or set of rules we can come up with that *is* principled that covers a good portion of the "remaining regressions" found in the wild.

> [requests for additional tests]

The last commit now adds fallback/nofallback revisions to several of the tests in tree to showcase the need for fallback. However, it also shows several tests that are not fixed by the current set of fallback rules (some pre-existing, some new in this PR):

* src/test/ui/never_type/fallback-closure-wrap.rs: This test requires us to consider an `as` cast for fallback, which is explicitly not part of fallback today - https://github.com/rust-lang/rust/blob/207d9558d00dd5cc438a6418ba96912d396e2155/compiler/rustc_typeck/src/check/mod.rs#L459-L460
* src/test/ui/never_type/defaulted-never-note.rs
   * This is a test that *could* be tackled with a simpler rule to the one we've added here, limited to just `?T: Foo` and `(): Foo`. IIRC, we need the more complicated rule because stabilization otherwise triggers errors in other tests. I think [binop-3](https://github.com/rust-lang/rust/blob/207d9558d00dd5cc438a6418ba96912d396e2155/src/test/ui/binop/binop-fail-3.rs) is a good example here. If we just have `?T: Foo` and `(): Foo`, then one side there will fallback to `()` (since it is `?T: PartialEq<_>`, but the other side has no such condition so results in `!` and `! == ()` is not implemented.
* src/test/ui/never_type/diverging-fallback-no-leak.rs
   * This seems like basically the same test as never-note, and I think is similarly explained as somewhat unavoidable.


@bors try @rust-timer queue

Rechecking performance after fixing our code to avoid testing `(): Foo` when we have a concrete type in the original predicate, but rather only for type vars. I don't think we can limit this to just *diverging* type vars, because those are only tracked when we detect a never to any adjustment -- and that may not have happened when we register this bound. This is basically the same reason we have to check these as they get registered rather than waiting until fallback and looking at the pending obligations.

(That said, the previous diff did not look terrible for what this code is doing, and I think we could accept the hit. I cannot readily explain the *improvements* -- I would expect the code added by this PR to mostly only make things worse).


---

I think the next steps here might be that we:

* Do several more rounds of review to get this in a "good enough" state to land, though likely not to stabilize never type
   * Would rather get this in-tree sooner rather than later to avoid rebasing it too much; it's gated behind feature gates anyway, so we don't need to decide on the specific rule set and get T-lang approval yet. (IMO, anyway).
* Open a PR stabilizing never type and collect a fresh round of crater data. My projection is ~32 regressions, with around 70 fixed by the last commit in this PR (based on comparison with the crater run with just Niko's commits).

**Comment by jackh726** [comment]
r=me here. I would feel better if @nikomatsakis can take a look and give the final signoff, but given that this is gated behind a feature gate, it might be better to merge and iterate in-tree.

**Comment by nikomatsakis** [comment]
> I am not sure that we will end up with "great" reasoning beyond the 'it fixes regressions'.

I think that is fine, I'd just like to comment on what kind of regression we are looking to fix.

**Comment by Mark-Simulacrum** [comment]
> > I am not sure that we will end up with "great" reasoning beyond the 'it fixes regressions'.
> 
> I think that is fine, I'd just like to comment on what kind of regression we are looking to fix.

There's now a comment in the fallback code pointing to the regression test the relationship code is fixing. Happy to edit it, or expand it - I think creating that repository for the alternatives documentation is a good idea - once I rebase the stabilization commits it'll be easier to off hand indicate which patterns break with particular algorithms without needing to refer to various hackmds :)

**Comment by jackh726** [comment]
@bors r=nikomatsakis,jackh726

even though it might be procedural, I'd like to note that I did review this (particularly because Niko did originally author many of these commits)

**Pull request merged**
